packages/typespec-vscode/src/npm-utils.ts

```diff
@@ -0,0 +1,143 @@
+
+export enum InstallationAction {
+  Install = "Install",
+  Upgrade = "Upgrade",
+  Skip = "Skip",
+  Cancel = "Cancel",
+}
+
+export enum npmDependencyType {
+  dependencies = "dependencies",
+  peerDependencies = "peerDependencies",
+  devDependencies = "devDependencies",
+}
+
+export interface NpmPackageInfo {
+  name: string;
+  version?: string;
+  resolved?: string;
+  overridden?: string;
+  dependencies?: Record<string, NpmPackageInfo>;
+}
+
+export class NpmUtil {
+  private cwd: string;
+
+  constructor(cwd: string) {
+    this.cwd = cwd;
+  }
+
+  public async npmInstallPackages(
+    packages: string[] = [],
+    options: any = {},
+    on?: spawnExecutionEvents,
+  ): Promise<ExecOutput> {
+    return spawnExecution("npm", ["install", ...packages], this.cwd, on);
+  }
+
+  public async ensureNpmPackageInstall(
+    packageName: string,
+    version?: string,
+  ): Promise<{ action: InstallationAction; version?: string }> {
+    const { installed: isPackageInstalled, version: installedVersion } =
+      await this.isPackageInstalled(packageName);
+    if (isPackageInstalled) {
+      if (version && installedVersion !== version) {
+        return {
+          action: InstallationAction.Upgrade,
+          version: version,
+        };
+      }
+      return { action: InstallationAction.Skip, version: installedVersion };
+    } else {
+      return { action: InstallationAction.Install, version: version };
+    }
+  }
+
+  public async ensureNpmPackageDependencyInstall(
+    packageName: string,
+    version?: string,
+    dependencyType: npmDependencyType = npmDependencyType.dependencies,
+    on?: spawnExecutionEvents,
+  ): Promise<string[]> {
+    const dependenciesToInstall: string[] = [];
+    let packageFullName = packageName;
+    if (version) {
+      packageFullName = `${packageName}@${version}`;
+    }
+
+    /* get dependencies. */
+    const dependenciesResult = await spawnExecution(
+      "npm",
+      ["view", packageFullName, dependencyType],
+      this.cwd,
+      on,
+    );
+
+    if (dependenciesResult.exitCode === 0) {
+      try {
+        // Remove all newline characters
+        let dependenciesJsonStr = dependenciesResult.stdout.trim();
+        dependenciesJsonStr = dependenciesJsonStr.replace(/\n/g, "");
+
+        // Change single quotes to double quotes
+        dependenciesJsonStr = dependenciesJsonStr.replace(/'/g, '"');
+        const json = JSON.parse(dependenciesJsonStr);
+        for (const [key, value] of Object.entries(json)) {
+          const { installed, version: installedVersion } = await this.isPackageInstalled(key);
+          if (installed && installedVersion) {
+            if (!this.isValidVersion(installedVersion, value as string)) {
+              dependenciesToInstall.push(`${key}@latest`);
+            }
+          }
+        }
+      } catch (err) {
+        if (on && on.onError) {
+          on.onError(err, "", "");
```

Comment:
I think we can log directly here, but if on... is used, please at least make sure the caller would log the error to make it easier to troubleshooting